### PR TITLE
Update lint declaration example

### LINKED ISSUE #712: emitting errors & diagnostics: example lint appears to be outdated
I was looking https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=LateContext#declaring-a-lint to find out how lints work and it seems the example is outdated, I could not find any references to `ExprWhile` in the rustc compiler code for example.

### PR DESCRIPTION
**Pull request by JohnTitor**
Fixes #712

### CODE REVIEW

**Comment by tshepang** [comment]
huh, such much speeds :open_mouth:

File: src/diagnostics.md [hunk 1]
```diff
@@ -300,35 +300,45 @@ declare_lint! {
...
+// This declares a struct and a lint pass, providing a list of associated lints. The
 // compiler currently doesn't use the associated lints directly (e.g., to not
 // run the pass or otherwise check that the pass emits the appropriate set of
 // lints). However, it's good to be accurate here as it's possible that we're
 // going to register the lints via the get_lints method on our lint pass (that
 // this macro generates).
-impl_lint_pass!(
-    WhileTrue => [WHILE_TRUE],
-);
+declare_lint_pass!(WhileTrue => [WHILE_TRUE]);
```
Comment (JohnTitor) [inline src/diagnostics.md:309]: We now use `declare_lint_pass` macro instead of defining manually. Should we do manually for clarify?

**Comment by JohnTitor** [comment]
Hm, the link error is unrelated to this, I'll submit another PR not to block the process.

**Comment by JohnTitor** [comment]
So, r? @nikomatsakis or @Mark-Simulacrum as per git blame

**Comment by Mark-Simulacrum** [comment]
Seems fine to me though without the context as to what upstream changes caused this I can't be sure the new code is good. Someone from clippy can likely give better guidance than I.

But r=me on the changes presuming they compile.

**Comment by JohnTitor** [comment]
I think the issue here is that `while_true` lint example is entirely outdated and now on `EarlyLintPass`.

> But r=me on the changes presuming they compile.

@Mark-Simulacrum It'd be great if you could _approve_ this so that "Review required" is satisfied :)

**Comment by JohnTitor** [comment]
cc @rust-lang/wg-rustc-dev-guide it'd be great if someone could also review here :)

**Comment by Mark-Simulacrum** [comment]
Oh sorry, missed the notification. Feel free to ping me on Zulip in the future, that's much more likely to get through.

**Comment by JohnTitor** [comment]
Thanks both!

> Feel free to ping me on Zulip in the future, that's much more likely to get through.

Gotcha!

**Pull request merged**